### PR TITLE
feat(types): support styling properties

### PR DESCRIPTION
**Pull request by sparten11740**
This is a proposal of how we can merge props that are defined inside the `componentStyle` function with the actual component props:

The below component would have all `Text` props available as well as the style props `{ transparent?: boolean; color: string }`
```ts
const StyledText = styled(Text)(({ transparent }: { transparent?: boolean; color: string }) => ({ flex: 1, opacity: transparent ? 0.5 : 1 }))
```

Depends on #20

### CODE REVIEW

File: test/index.type-test.tsx [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+      <StyledText color="red" transparent>Hello Transparent World</StyledText>
+      {/* should also work without transparent prop */}
+      <StyledText color="red">Hello World</StyledText>
+      {/* @ts-expect-error -- should not work without required color prop */}
+      <StyledText>Hello World</StyledText>
+      {/* @ts-expect-error -- big is only available on StyledScalableText */}
+      <StyledText color="red" big>Hello World</StyledText>
+      <StyledScalableText color="red" transparent big>Hi</StyledScalableText>
+      {/* @ts-expect-error -- transparent prop is not allowed */}
+      <StyledTextWithObjectProps transparent>Hi</StyledTextWithObjectProps>
```
Comment (sparten11740) [inline test/index.type-test.tsx:23]: note these type tests, they are a useful development tool but also help to prevent regressions

Comment (headfire94) [reply test/index.type-test.tsx:23]: how we avoid passing transaprent to actual native component?

Comment (headfire94) [reply test/index.type-test.tsx:23]: ah didn't notice comment, all good

Comment (sparten11740) [reply test/index.type-test.tsx:23]: this is just typing with no runtime effects and supports the previous usage patterns that we have
> The reason is that with shakl with pass this custom prop down to component - this is not expected. right now react-native allows that

it doesn't change anything about that

**Comment by headfire94** [comment]
thanks! i'll extend type-test also with other cases an let you know if any other help needed!

File: test/index.type-test.tsx [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import React, { Text } from 'react-native'
+import styled from '../src/index'
+
+const StyledText = styled(Text)(({ transparent }: { transparent?: boolean; color: string }) => ({ flex: 1, opacity: transparent ? 0.5 : 1 }))
```
Comment (headfire94) [inline test/index.type-test.tsx:5]: @sparten11740 will it work with `styled.Text` ?

Comment (sparten11740) [reply test/index.type-test.tsx:5]: we don't currently have the types for that defined, do we?

Comment (sparten11740) [reply test/index.type-test.tsx:5]: ah overlooked them, they are in a different file

Comment (headfire94) [reply test/index.type-test.tsx:5]: i was thinking it's here

https://github.com/ExodusMovement/shakl/blob/befcafb83a22656d188c8715ead0ddbf3be2ee24/src/rn.tsx#L27

Comment (headfire94) [reply test/index.type-test.tsx:5]: for usage https://github.com/ExodusMovement/shakl/blob/befcafb83a22656d188c8715ead0ddbf3be2ee24/src/rn.tsx#L62

Comment (sparten11740) [reply test/index.type-test.tsx:5]: I am a little confused by the type, it specifies we have `View, Text, Image, Touchable, TextInput, ScrollView` but we only define `View, Text, Image, Touchable`. what am I missing?

Comment (headfire94) [reply test/index.type-test.tsx:5]: actually i need to remove TextInput and ScrollView to not introduce too many changes in PR. but ideally add them too

Comment (sparten11740) [reply test/index.type-test.tsx:5]: https://github.com/ExodusMovement/shakl/pull/21/commits/1480b12b095cf07279f9d9ac1f6ea0b08f4bce2e how the extended styles could be dealt with

Comment (sparten11740) [reply test/index.type-test.tsx:5]: we don't need the explicit types besides passing in the generics to the styled function

**Pull request merged**